File: src/Reloaded.Memory.Buffers/Utilities/Mathematics.cs
namespace Reloaded.Memory.Buffers.Utilities;

internal static class Mathematics
{
    /// <summary>
    /// Rounds up a specified number to the next multiple of X.
    /// </summary>
    /// <param name="number">The number to round up.</param>
    /// <param name="multiple">The multiple the number should be rounded to.</param>
    /// <returns></returns>
    internal static nuint RoundUp(nuint number, nuint multiple)
    {
        if (multiple == 0)
            return number;

        nuint remainder = number % multiple;
        if (remainder == 0)
            return number;

        return number + multiple - remainder;
    }

    /// <summary>
    /// Rounds up a specified number to the previous multiple of X.
    /// </summary>
    /// <param name="number">The number to round down.</param>
    /// <param name="multiple">The multiple the number should be rounded to.</param>
    internal static nuint RoundDown(nuint number, nuint multiple)
    {
        if (multiple == 0)
            return number;

        nuint remainder = number % multiple;
        if (remainder == 0)
            return number;

        return number - remainder;
    }

    /// <summary>
    /// Rounds up a specified number to the previous multiple of X.
    /// </summary>
    /// <param name="number">The number to round down.</param>
    /// <param name="multiple">The multiple the number should be rounded to.</param>
    /// <returns></returns>
    internal static nuint RoundDown(nuint number, int multiple) => RoundDown(number, (nuint)multiple);

    /// <summary>
    /// Rounds up a specified number to the next multiple of X.
    /// </summary>
    /// <param name="number">The number to round up.</param>
    /// <param name="multiple">The multiple the number should be rounded to.</param>
    internal static nuint RoundUp(nuint number, int multiple) => RoundUp(number, (nuint)multiple);

    /// <summary>
    /// Returns smaller of the two values.
    /// </summary>
    /// <param name="a">First value.</param>
    /// <param name="b">Second value.</param>
    internal static nuint Min(nuint a, nuint b) => a < b ? a : b;

    /// <summary>
    /// Adds the two values, but caps the result at MaxValue if it overflows.
    /// </summary>
    /// <param name="a">First value.</param>
    /// <param name="b">Second value.</param>
    internal static nuint AddWithOverflowCap(nuint a, nuint b)
    {
        var max = unchecked((nuint)(-1));
        if (max - a >= b)
            return a + b;

        return max;
    }

    /// <summary>
    /// Subtracts the two values, but caps the result at MinValue if it overflows.
    /// </summary>
    /// <param name="a">First value.</param>
    /// <param name="b">Second value.</param>
    public static nuint SubtractWithUnderflowCap(nuint a, nuint b) => b <= a ? a - b : 0;
}

File: src-rust/src/utilities/mathematics.rs
/// Rounds up a specified number to the next multiple of X.
///
/// # Arguments
///
/// * `number` - The number to round up.
/// * `multiple` - The multiple the number should be rounded to.
pub fn round_up<T: Into<usize>>(number: usize, multiple: T) -> usize {
    let multiple = multiple.into();
    if multiple == 0 {
        return number;
    }

    let remainder = number % multiple;
    if remainder == 0 {
        return number;
    }

    number + multiple - remainder
}

/// Rounds down a specified number to the previous multiple of X.
///
/// # Arguments
///
/// * `number` - The number to round down.
/// * `multiple` - The multiple the number should be rounded to.
pub fn round_down<T: Into<usize>>(number: usize, multiple: T) -> usize {
    let multiple = multiple.into();
    if multiple == 0 {
        return number;
    }

    let remainder = number % multiple;
    if remainder == 0 {
        return number;
    }

    number - remainder
}

/// Returns smaller of the two values.
///
/// # Arguments
///
/// * `a` - First value.
/// * `b` - Second value.
#[allow(dead_code)]
pub fn min(a: usize, b: usize) -> usize {
    a.min(b)
}

/// Adds the two values, but caps the result at MaxValue if it overflows.
///
/// # Arguments
///
/// * `a` - First value.
/// * `b` - Second value.
pub fn add_with_overflow_cap(a: usize, b: usize) -> usize {
    a.saturating_add(b)
}

/// Subtracts the two values, but caps the result at MinValue if it overflows.
///
/// # Arguments
///
/// * `a` - First value.
/// * `b` - Second value.
pub fn subtract_with_underflow_cap(a: usize, b: usize) -> usize {
    a.saturating_sub(b)
}
